### PR TITLE
Use globalization for media stream language names

### PR DESCRIPTION
**Pull request by bilde2910**
This change request concerns how Jellyfin translates ISO 639-2 language codes (e.g. `eng`) to display names (e.g. `English`) when displaying media info. Jellyfin's current behavior as of 10.7.z is to use [CultureInfo](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.threeletterisolanguagename?view=net-5.0) to resolve language names (first committed in #2773 and merged in #2788). The proposed change is to use Jellyfin's globalization infrastructure to handle this instead, and only using CultureInfo as a fallback when an unknown lanugage is encountered.

The reasons for this change request is as follows:

- CultureInfo supplies a `DisplayName` that appears to be locked to English. If a user changes their language in Jellyfin, they will still see English language names in the audio and subtitle selection boxes when viewing media in Jellyfin.
- CultureInfo does not contain mappings for all major languages. E.g. `swe` is mapped to `Swedish`, `dan` to Danish, but `nor` does not have a mapping, and falls back to displaying "Nor" in Jellyfin.
- CultureInfo only recognizes ISO 639-2/T codes, resulting in no mapping for media that indicates an ISO 639-2/B code that differs from its T code, e.g. French (`fre` (B) vs `fra` (T)) and German (`ger` (B) vs `deu` (T)).

Using Jellyfin's globalization infrastructure allows to overcome all of these shortcomings of using CultureInfo. However, this approach also has some shortcomings:

- If any languages are covered by .NET CultureInfo, but matching keys do not exist in localization files, mapping will fail
- An increased number of strings to localize for translators. The current proposed set of languages to cover will add 185 new strings to Jellyfin core.

To work around these shortcomings, I propose keeping the CultureInfo lookup mechanism as a fallback. If a language exists in Jellyfin globalization files, that value will be used for rendering the full language name. If it does not exist, Jellyfin will fall back to CultureInfo. If CultureInfo also fails, fall back to displaying the language code instead, as was the default in Jellyfin up to version 10.6.z.

Regarding the number of extra strings, I am willing to help out by importing the language names from some of the most commonly used languages from the ISO 639-1 tables on Wikipedia in Weblate. Most languages should have official lists of language names somewhere that could also possibly be used to aid this process.

The goal of this change request is to improve user experience regarding language selection for media in Jellyfin, making it more reliable than it currently is in English, and improving user experience by rendering language names in the user's native language for all other Jellyfin locales (where available).

## Technical changes

- `LanguageXxx` keys have been added to globalization files where `Xxx` is the ISO 639-2/B code for the language in question. 185 such keys have been added.
- The prefix of `Language` was used to avoid possible future conflicts, and to clearly group these together and indicate their purpose.
- `SqliteItemRepository.cs` and `ProbeResultNormalizer.cs` will look up this key using the `Language` attribute of the `MediaStream` upon creation, and either set the `localizedLanguage` attribute of MediaStream to the localized name, or to null, depending on whether or not the language exists in globalization files.
- `MediaStream` will use `localizedLanguage` if it is non-null, and CultureInfo if null (fallback).

## Relevant issues/references

- CultureInfo commited in #2773
- Merged in #2788
- Implements FR [534](https://features.jellyfin.org/posts/534/subtitle-abbreviation)

### CODE REVIEW

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 1]
```diff
@@ -6067,6 +6068,13 @@ private MediaStream GetMediaStream(IReadOnlyList<IResultSetValue> reader)
             if (reader[4].SQLiteType != SQLiteType.Null)
             {
                 item.Language = reader[4].ToString();
+
+                string i18nKey = "Language" + StringHelper.FirstToUpper(item.Language);
```
Comment (crobibero) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs:6072]:
```suggestion
                string i18nKey = "Language" + item.Language;
```

The localized string dictionary is case insensitive so this isn't needed

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs [hunk 2]
```diff
@@ -650,6 +651,13 @@ private MediaStream GetMediaStream(bool isAudio, MediaStreamInfo streamInfo, Med
                 stream.Language = GetDictionaryValue(streamInfo.Tags, "language");
                 stream.Comment = GetDictionaryValue(streamInfo.Tags, "comment");
                 stream.Title = GetDictionaryValue(streamInfo.Tags, "title");
+
+                string i18nKey = "Language" + StringHelper.FirstToUpper(stream.Language);
```
Comment (crobibero) [inline MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:655]:
```suggestion
                string i18nKey = "Language" + stream.Language;
```

same

**Comment by bilde2910** [comment]
Thanks @crobibero! That simplifies the code a little. Just committed the no longer necessary `using` as well.

**Review by BaronGreenback — APPROVED** [review]

**Review by crobibero — DISMISSED** [review]

**Comment by Smith00101010** [comment]
This is now approved for quite a while and I would love to have this feature. Is there a reason why this isn't merged yet? Or any way I can help to get it merged?

**Comment by cvium** [comment]
Would this fulfill the same purpose https://github.com/ForeverZer0/Iso639?

**Comment by bilde2910** [comment]
From what I can tell, Nager.Country does not cover ISO 639-2/B codes which are common in media streams. Iso639 does, but it appears to be very expansive, much more than what would be really needed by Jellyfin - it has over 7800 languages, including ancient, extinct and extremely rarely used ones, which would be a nightmare to localize to other languages for Jellyfin.

Nager.Country offers translations, Iso639 does not. Overall I feel that this kind of functionality is simple enough in itself that it's better to code it into Jellyfin rather than relying on an external dependency for it. My implementation is just 11 lines of code excluding lines with only brackets and the actual localized data.

**Comment by EraYaN** [comment]
This implementation does ad a ton of language strings though. And is does not really do the same "here is a list of everything" that for example Iso639 does (with the possibility of translating between 3 and 2 latter codes). And are translations really required? If the language name is available in English (and maybe the language itself) that should be enough right, I think people should be able to recognize their own language in English.

So we could probably use both to works with both 3 letter language codes (from Iso639 to convert to 2 letter) and translations from Nager.Country (using the 2 letter language codes)

**Comment by bilde2910** [comment]
It does add a bunch of language strings, but as noted in the PR, I fully intend to help out importing language names into Weblate as soon as this is merged, to reduce the burden on other translators. The feature request had someone request that language names be translated as well, and it's something that would improve the user experience of Jellyfin for non-English speakers.

Furthermore, Iso639, or rather the underlying base data from SIL, has a license that prohibits extension of the language list. I have already encountered several instances of a non-standard language code being used in metadata on retail movie discs that I have in my library. If we use SIL's list, we would never be able to add those into Jellyfin's set of language codes. Having it be part of Jellyfin itself would be more flexible in that regard.

**Comment by EraYaN** [comment]
I mean the point of the SIL list is that it's by definition the correct one for the ISO standard. So anything is is not a language. So to me that sounds like a feature.

Any weird creative choice with language codes and country codes is just wrong and those disk are probably not conform their respective specification anyway. (And to keep those out is also why SIL is so picky about it and IMO they are right.)

And we would still violate their Terms of Use if we put the languages internally anyway. Since it seems the code tables themselves are covered.

**Comment by bilde2910** [comment]
We can still put a code table inside Jellyfin, but it cannot be sourced from SIL; we need to get the language codes from elsewhere (and they are well documented outside of SIL's list). Regarding non-standard language codes, I figured out the distinction here. BluRay appears to use ISO 639-2/B. SIL's list is ISO 639-3. As a side note, I [found a list](https://forum.doom9.org/showthread.php?t=155762) of language codes used by BluRay. At least one of the languages in there are defined in 639-2, but not in 639-3 because they are language groups, not individual languages. Thus they are still valid, but outdated and removed from the latest spec (but still in use by BluRay).

**Comment by bilde2910** [comment]
Hi again! Just checking in and asking if there's anything more that I should do before this may be merged, seeing as this PR has been in the approved state for a few months now. :-)

**Comment by EraYaN** [comment]
The hesitancy to add all these language strings still exists, I guess.

**Comment by bilde2910** [comment]
I suppose. I will however reiterate that I will personally add as many localizations of those strings as possible, to avoid that burden falling on the other volunteer translators. I was planning to do so via Weblate, but I can also commit the translations directly to this PR prior to merging if that would be preferable. My hesitance of external dependencies would in part be ensuring that whatever language we choose works, that it has a compatible license, and most of all, that it has full 639-2/B coverage.

**Comment by cvium** [comment]
Seems like we've forgotten about this again. I'm not sure much has changed since https://github.com/jellyfin/jellyfin/pull/4679#issuecomment-823631986 though.

**Comment by bilde2910** [comment]
I'm working on pre-importing all the localized language strings to this PR now as opposed to adding them to Weblate later. This is so there will be as little as possible that others have to do in Weblate. I'm searching around for official names of languages in all the other languages right now.

I still maintain that translating the language names will result in a better user experience, which is why I'm doing this, and recommending it over another package-based approach that may be incomplete in terms of language codes, that we can't extend if needed, and that's primarily locked to English names only.

**Comment by bilde2910** [comment]
I've imported a bunch of languages from Wikipedia, but only those that I've been able to verify (as I can't read non-latin scripts).

Current coverage of language names:

    de.json    : 185/185 (100%)
    nl.json    : 185/185 (100%)
    es.json    : 185/185 (100%)
    fr.json    : 185/185 (100%)
    en-US.json : 185/185 (100%)
    eo.json    : 184/185 (99%)
    it.json    : 184/185 (99%)
    pt.json    : 184/185 (99%)
    pl.json    : 184/185 (99%)
    fi.json    : 183/185 (98%)
    sv.json    : 182/185 (98%)
    gl.json    : 181/185 (97%)
    ca.json    : 180/185 (97%)
    nb.json    : 180/185 (97%)
    af.json    : 177/185 (95%)
    lt-LT.json : 176/185 (95%)
    cs.json    : 174/185 (94%)
    id.json    : 174/185 (94%)
    nn.json    : 173/185 (93%)
    tr.json    : 169/185 (91%)
    hr.json    : 163/185 (88%)
    lv.json    : 161/185 (87%)
    bg-BG.json : 160/185 (86%)
    ms.json    : 144/185 (77%)
    is.json    : 143/185 (77%)
    ro.json    : 143/185 (77%)
    da.json    : 139/185 (75%)
    hu.json    : 138/185 (74%)
    sk.json    : 126/185 (68%)
    sl-SI.json : 96/185 (51%)

**Comment by WWWesten** [comment]
Translating every language name into every language is a nightmare. It should be done like, for example, the list of languages on Wikipedia. The name of each language should be given in that very language. Like:
`"LanguageGer": "Deutsch",`
`"LanguageHeb": "עִבְרִית",`
`"LanguagePer": "فارسی",`
`"LanguageRus": "Русский",`

Remove extinct, not spoken languages, languages with <100000 speakers, minor dialekts etc.

**Comment by bilde2910** [comment]
> Translating every language name into every language is a nightmare.

Much of the ground work for this is already established, and the work already >95% complete for many of the most commonly spoken languages. I've been looking up official names of languages from several countries' government websites to try to fill in gaps as well. I've been busy lately, but I plan to continue soon. I will do this work alone if I have to.

> It should be done like, for example, the list of languages on Wikipedia. The name of each language should be given in that very language. Like: `"LanguageGer": "Deutsch",` `"LanguageHeb": "עִבְרִית",` `"LanguagePer": "فارسی",` `"LanguageRus": "Русский",`

Disagree, because the user experience of having language names translated to the user's own language is significantly better than having each language name localized only to their own official names. A particular example for this is where a user may want to watch content that is available in e.g. Japanese and Chinese, the user wants to listen to the content in the original non-dubbed language of the content with English subtitles. Differentiating between 日本 and 中国人 is impossible for a user that can't read either language and haven't seen them before. It also makes it more clear which languages are available for the media in question in Jellyfin when the user doesn't speak all languages.

> Remove extinct, not spoken languages

Agree; this should be done, some may have inadvertently made it onto the list. If I come across them, they will be removed.

> languages with <100000 speakers, minor dialekts etc.

Disagree; there are many languages with few speakers that have names that are very easily translatable to other languages. Personally I have multiple entries in my Jellyfin instance that are in a language with ~25k speakers, and would prefer that those are treated equally to the more commonly spoken ones. There should by all means be a threshold, but it must be much lower.

**Comment by bilde2910** [comment]
This PR has been marked "discussion needed" for a while, and has passed its birthday mark by now. I know you all are really busy, and I don't mean to interrupt, but any thoughts on this one @crobibero? :-)

**Comment by crobibero** [comment]
I don't believe anything has changed since the last post.

Additionally, I think that the subtitle language should not be translated. It should be safe to assume that the user enabling the subtitle can read the language they want

**Comment by joshuaboniface** [comment]
@jellyfin/core @jellyfin/server Any remaining thoughts on this?

**Comment by EraYaN** [comment]
I still think it's a misuse of the localization system when there are perfectly good official lists and libraries out there with all the translation options in there.

**Comment by Bond-009** [comment]
Closing this as it has been almost 2 years. Discussions on the correct way forward can happen on GH Discussions, Matrix or the forum

**Pull request closed without merging**